### DISCUSSION TITLE
How to set numbering only for specific heading level?

### DISCUSSION THREAD

**Discussion by creatorofuniverses** [typst/typst#1363] [Q&A]
Let's say I want to be numbered only in the headers of level 1, and the rest went without numbering (those no 1.1, 1.a, and so on). How can I do this?

The idea is to redefine the function, but I don't really understand how to do it correctly.

P.S. It seems like it would be great to add this to the functionality of the original function, but that's a separate topic

**Comment by lvignoli** [comment]
This works, but it feels really like a hack, since one has to redefine the show rules by hand for all the headings:

```typst
#set heading(numbering: "1.1.")
#show heading: it => block(it.body)
#show heading.where(level: 1): it => block(it.numbering.at(0) + ". " + it.body)

= Foo
== Bar
```

Maybe numbering patterns handle that better, say `1._` to number only level 1, or `_.1` to number all but level 1.

**Comment by richardkraus-adesso** [comment]
You could use [numbering functions](https://typst.app/docs/reference/meta/numbering/) like this:
```typst
#set heading(numbering: (first, ..other) =>
  if other.pos().len() == 0 { return first }
)

= This is numbered
== This is not
```
![image](https://github.com/typst/typst/assets/136313436/61d84fa9-8501-4386-89d2-d6bb6baa9ce4)

imo it would be nicer to have some way of modifying the style of the current element, similar to how [show-set rules](https://typst.app/docs/reference/styling/#show-rules) are written but also affecting itself

**Comment by lvignoli** [comment]
@richardkraus-adesso thanks that's much tidier 😀

I know there is some "style" work under the hood, this might provides a way to do so.

**Comment by rem3-1415926** [comment]
I think the solution of @richardkraus-adesso can be generalised to support any heading depth: 
```
#let level = 2
#set heading(
  numbering: (..numbers) =>
    if numbers.pos().len() <= level {
      return numbering("1.a.", ..numbers)
    }
)
= Numbered header
== Numbered subheader
=== non-numbered subsubheader
```
![2023-08-21_11-37](https://github.com/typst/typst/assets/43850766/34055148-e7db-4f57-a022-3f3926221b78)
Passing the `numbers` arguments back to the numbering function returns the formatting work to Typst.

**Reply by richardkraus-adesso** [reply]
Using functions, you can also make that more readable.
What I ended up using in my case was along the lines of this:
```typst
#let number-until-with(max-level, schema) = (..numbers) => {
  if numbers.pos().len() <= max-level {
    numbering(schema, ..numbers)
  }
}

#set heading(numbering: number-until-with(2, "1.a"))

= Numbered Header

== Numbered subheader

=== non-numbered subheader
```
By defining a function you can reuse the numbering function anywhere and move the implementation outside the `#set` call

**Comment by th7mo** [comment]
@VitalyyBezuglyj just use the `where()` filter to only target level 1 headings:

```typst
#show heading.where(level: 1): set heading(numbering: "1")
```

I don't know if this works with a TOC tho...

**Comment by ameresv** [comment]
Hi Guys! This is more a question than an answer. I came around to this topic. 
How can I approach the scenario where I want to start the numbering from H2 and so on, based on a "1." format? 
I mean:

# Not numbered
## 1 Numbered "(1.0 or just 1)"
### 1.1 Numered "(1.1)"

Regards

**Reply by lvignoli** [reply]
One way to go is 

```typst
#set heading(numbering: (..nums) => {
  if nums.pos().len() == 1 { return none }
  nums.pos().slice(1).map(str).join(".") + "."
  }
)
```

There is small alignement issue for level 1 heading, as it seems despite having a `none` content for the numbering, default typst inserts the space between the numbering and the heading title.
Instead of returning `none`, you can return a negative space, such as `h(-0.5em)`, but it's brittle and must be tuned to your font, etc.

**Reply by lvignoli** [reply]
![Capture d’écran 2025-05-28 à 10 22 24](https://github.com/user-attachments/assets/9d0d032e-7fde-4d77-a8cb-4251f88678da)

**Reply by ameresv** [reply]
Thanks for the answer. Looks very complex for a newbie in typst like me. 
Can you help me convert this function into a #show applied to these headings? 
Warm regards.

**Reply by ShortDevelopment** [reply]
```typst
// Set numbering to `1.` but ignore the first level
#set heading(numbering: (first, ..nums) => numbering("1.", ..nums))
  
// Remove the space in front of headings(level: 1)
#show heading.where(level: 1): set heading(numbering: none)

= This is not numbered
== This is
== This is
=== This is
  ```
